tools/splat_ext/pm_sbn.py

```diff
@@ -0,0 +1,738 @@
+        self.entriesSize, = struct.unpack(">h", data[0x0A:0x0C])
+        self.tblOffset, = struct.unpack(">h", data[0x0C:0x0E])
+        self.tblSize, = struct.unpack(">h", data[0x0E:0x10])
+        self.shortsOffset, = struct.unpack(">h", data[0x10:0x12])
+        self.shortsSize, = struct.unpack(">h", data[0x12:0x14])
+        self.unk_14 = data[0x14:0x20]
+        assert(self.signature == b"INIT")
+        return 0x20
+
+
+# see au_load_BK_headers
+class BufferEntry:
+    fileIndex: int
+    bankIndex: int
+    bankGroup: int
+
+    def decode(self, data: bytes) -> int:
+        self.fileIndex, = struct.unpack(">h", data[0x00:0x02])
+        self.bankIndex = data[0x02]
+        self.bankGroup = data[0x03]
+        return 0x04
+
+    def __str__(self) -> str:
+        return f"BufferEntry({self.fileIndex}, {self.bankIndex}, {self.bankGroup})"
+
+    def __hash__(self) -> int:
+        return hash((self.fileIndex, self.bankIndex, self.bankGroup))
+
+
+class EndOfDataException(Exception):
+    pass
+
+
+class InitSongEntry:
+    bgmFileIndex: int
+    bkFileIndex: List[int]
+
+    def decode(self, data: bytes) -> int:
+        self.bgmFileIndex, = struct.unpack(">h", data[0x00:0x02])
+        if self.bgmFileIndex < 0:
+            raise EndOfDataException()
+        self.bkFileIndex = [
+            struct.unpack(">h", data[0x02:0x04])[0],
+            struct.unpack(">h", data[0x04:0x06])[0],
+            struct.unpack(">h", data[0x06:0x08])[0]
+        ]
+        return 0x08
+
+    def __str__(self) -> str:
+        return f"InitSongEntry({self.bgmFileIndex}, {self.bkFileIndex})"
+
+    def __hash__(self) -> int:
+        return hash((self.bgmFileIndex, sum(self.bkFileIndex)))
+
+
+if splat_loaded:
+    class N64SegPm_sbn(N64Segment):
+        def split(self, rom_bytes):
+            dir = options.opts.asset_path / self.dir / self.name
+            data = rom_bytes[self.rom_start : self.rom_end]
+
+            if dir.exists():
+                raise Exception(f"SBN directory {dir} already exists")
+
+            sbn = SBN()
+            byte_count = sbn.decode(data)
+            sbn.write(dir)
+
+            if self.rom_end:
+                assert byte_count == self.rom_end - self.rom_start
+            else:
+                self.rom_end = self.rom_start + byte_count
+
+        def get_linker_entries(self):
+            dir = options.opts.asset_path / self.dir / self.name
+            out = options.opts.asset_path / self.dir / (self.name + ".sbn")
+
+            sbn = SBN()
+            config_files = sbn.read(dir) # TODO: LayeredFS/AssetsFS read, supporting merges
+            inputs = config_files + [dir / f.file_name() for f in sbn.files]
+            return [
+                LinkerEntry(
+                    self,
+                    inputs,
+                    out,
+                    ".data",
+                ),
+            ]
+
+def get_song_symbol_name(song_id: int) -> str:
+    song_names = {
```

Comment:
Shouldn't this be in a separate yaml file?

Reply:
yeah, probably... and separated by version. so like ver/us/song-names.yaml